tests/python/pants_test/engine/test_rules.py

```diff
@@ -389,13 +410,16 @@ def d_for_b(b: B) -> D:
     with self.assertRaises(Exception) as cm:
       create_scheduler(rules)
 
-    self.assert_equal_with_printing(dedent("""
+    self.assert_equal_with_printing(
+      dedent(
+        """\
         Rules with errors: 1
-          (D, [B], d_for_b()):
-            Was not usable by any other @rule.
-        """).strip(),
-        str(cm.exception)
-      )
+          Rule(func=d_for_b(), product=D, params=[B]):
+            Was not reachable, possibly because it was shadowed by another rule.
```

Comment:
```suggestion
            Was not reachable, possibly because no rules can produce B, or it was shadowed by another rule.
```

Reply:
Happy to punt, but I think what you want for an intrinsic is something like: `format!("{}", intrinsic.input)` - if it would help to add a:
```rust
trait HasInputTypes {
  fn input_types(&self) -> Vec<TypeId>;
}
```
and implement it for both `Task` and `Intrinsic`, we can do that :)

Reply:
Makes sense. I'm going to punt on this. The followup is linked to in https://github.com/pantsbuild/pants/issues/7509#issuecomment-577445094.